Range cleanup and support range with two CLR types

roji opened the pull request:
The first commit cleans up so messiness around the design of the range handler, making it inherit directly from NpgsqlTypeHandler<T> and removing lots of needlessly duplicated code.

The second commit makes it possible for the range handler to handle more than one CLR type, and adds support for `NpgsqlRange<DateTimeOffset>`, fixing #2436.

roji commented:
/cc @Brar, you've been looking at ranges recently (although this is a different concern).

YohDeadfall commented on src/Npgsql/TypeHandlers/RangeHandler.cs:
```diff
@@ -181,4 +154,35 @@ public async Task Write(NpgsqlRange<TElement> value, NpgsqlWriteBuffer buf, Npgs
         #endregion
     }
+
+    interface IRangeHandlerWithTwoElements
```
Looks ugly.

Brar replied:
I kind of agree.
The ugliness lies in the fact that - as I understand it - it doesn't really open up the handler to support any base type for ranges.

What if I opened an issue to support NpgsqlRange\<LocalDateTime> in the NodaTime plugin?

Would we then make it a IRangeHandlerWithThreeElements or would we just write an new handler for the plugin and hook that one up instead of the normal one?

Brar replied:
Why not change the interface to following (the `fieldDescription` is superfluous in this interface anyways):

```csharp
public interface IRangeHandlerWithMultipleElements
{
    IEnumerable<Type> AdditionalFieldTypes { get; }
}
```

Then we could do range handlers that support as many types as we deem necessary like the following:
```csharp
public class RangeHandler<TElement1, TElement2> : RangeHandler<TElement1>, INpgsqlTypeHandler<NpgsqlRange<TElement2>>,
    IRangeHandlerWithMultipleElements
{
    public virtual IEnumerable<Type> AdditionalFieldTypes
    {
        get { yield return typeof(NpgsqlRange<TElement2>); }
    }

    public RangeHandler(NpgsqlTypeHandler elementHandler) : base(elementHandler) {}

    ValueTask<NpgsqlRange<TElement2>> INpgsqlTypeHandler<NpgsqlRange<TElement2>>.Read(NpgsqlReadBuffer buf, int len, bool async, FieldDescription fieldDescription)
        => DoRead<TElement2>(buf, len, async, fieldDescription);

    public int ValidateAndGetLength(NpgsqlRange<TElement2> value, ref NpgsqlLengthCache lengthCache, NpgsqlParameter parameter)
        => ValidateAndGetLength<TElement2>(value, ref lengthCache, parameter);

    public Task Write(NpgsqlRange<TElement2> value, NpgsqlWriteBuffer buf, NpgsqlLengthCache lengthCache, NpgsqlParameter parameter, bool async)
        => Write<TElement2>(value, buf, lengthCache, parameter, async);
}

public class RangeHandler<TElement1, TElement2, TElement3> : RangeHandler<TElement1, TElement2>, INpgsqlTypeHandler<NpgsqlRange<TElement3>>,
    IRangeHandlerWithMultipleElements
{
    public override IEnumerable<Type> AdditionalFieldTypes
    {
        get
        {
            yield return typeof(NpgsqlRange<TElement2>);
            yield return typeof(NpgsqlRange<TElement3>);
        }
    }

    public RangeHandler(NpgsqlTypeHandler elementHandler) : base(elementHandler) {}

    ValueTask<NpgsqlRange<TElement3>> INpgsqlTypeHandler<NpgsqlRange<TElement3>>.Read(NpgsqlReadBuffer buf, int len, bool async, FieldDescription fieldDescription)
        => DoRead<TElement3>(buf, len, async, fieldDescription);

    public int ValidateAndGetLength(NpgsqlRange<TElement3> value, ref NpgsqlLengthCache lengthCache, NpgsqlParameter parameter)
        => ValidateAndGetLength<TElement3>(value, ref lengthCache, parameter);

    public Task Write(NpgsqlRange<TElement3> value, NpgsqlWriteBuffer buf, NpgsqlLengthCache lengthCache, NpgsqlParameter parameter, bool async)
        => Write<TElement3>(value, buf, lengthCache, parameter, async);
}
```

roji replied:
You guys are right, I did this a bit quickly without thinking more generally.

I've just pushed a commit that follows much of what @Brar proposed - take a look. Since our type handler system is generics-based, if we want a range handler for 3 types we need a new RangeHandler class (much like there are lots of ValueTuple classes with different numbers of type parameters). I don't think there's a way around that.

I also fixed an important problem with the previous implementation - we were binding the second range CLR type without any sort of check in element's handler's bound CLR types: that's incorrect. The idea is that the handlers themselves don't decide which CLR types they bind to - the user has control over that. So we now filter the range handler's exposed list via its element handler's list, and only bind range types whose corresponding element types have been bound. Hope that's clear, it's definitely a very elaborate type handling model :)

The fieldDescription is basically there for BitStringHandler, which returns a different type based on the type modifier (a single bool for one, or a BitArray otherwise). So in theory we don't support a range that will automatically be over bool or BitArray based on the fieldDescription - I think that's OK :)

Brar replied:
> Since our type handler system is generics-based, if we want a range handler for 3 types we need a new RangeHandler class (much like there are lots of ValueTuple classes with different numbers of type parameters). I don't think there's a way around that.

That's a pretty common scenario. Also see `Action`, `Func` and others.
I don't think that we will bind like hundreds of types to `RangeHandler` so that seems absolutely ok to me.

> So we now filter the range handler's exposed list via its element handler's list, and only bind range types whose corresponding element types have been bound.

Makes sense.

> So in theory we don't support a range that will automatically be over bool or BitArray based on the fieldDescription - I think that's OK :)

I wouldn't know what to do with such a beast 😜

YohDeadfall commented on src/Npgsql/TypeHandlers/RangeHandler.cs:
```diff
@@ -9,12 +9,6 @@
 
 namespace Npgsql.TypeHandlers
 {
-    public abstract class RangeHandler : NpgsqlTypeHandler
```
Any reason to remove it? Why not to add an abstract method to get all available types?

roji replied:
One of the main points of this PR is to remove all the duplicated code in `RangeHandler<TElement>` that was copied as-is from `NpgsqlTypeHandler<T>` (like `Read<TAny>`, `ValidateAndGetLength<TAny>`...). The only reason these couldn't simply be inherited directly from `NpgsqlTypeHandler<T>` (like other type handlers) is that we have the non-generic RangeHandler as a superclass, adding no value at all (I really don't know what I was thinking...).

YohDeadfall replied:
The current trend is to have abstract classes instead interfaces if there is no multiple inheritance. By the way, if I understand Core CLR documentation correctly it says that interface calls are slower than instance calls ([source](https://github.com/dotnet/coreclr/blob/master/Documentation/botr/virtual-stub-dispatch.md#current-state)):

> Currently, VSD is enabled only for interface method calls but not virtual instance method calls. There were several reasons for this:
>
> * **Startup:** Startup working set and speed were hindered because of the need to generate a great deal of initial stubs.
> * **Throughput:** While interface dispatches are generally faster with VSD, virtual instance method calls suffer an unacceptable speed degradation
>
> As a result of disabling VSD for virtual instance method calls, every type has a vtable for virtual instance methods and the implementation table described above is disabled. Dispatch maps are still present to enable interface method dispatching.

roji replied:
@YohDeadfall but as I wrote above, the point was to remove the duplication of code from NpgsqlTypeHandler<T> into RangeHandler<T>, which was the case before when RangeHandler<T> inherited from a non-generic RangeHandler.

Regarding perf, this is one of those places where I think we should prefer design correctness and codebase quality over such a micro-optimization. Even if there's a small degradation associated with interface dispatch, this could be a temporary situation and overall I doubt that it'd be decisive.

roji commented:
Guys, thanks for your comments - please re-review.

Brar commented on src/Npgsql/TypeHandlers/RangeHandler.cs:
```diff
@@ -17,15 +22,23 @@ namespace Npgsql.TypeHandlers
 
+        public Type[] FieldTypes { get; }
+
+        public RangeHandler(NpgsqlTypeHandler elementHandler, Type[] fieldTypes)
```
This constructor should rather be protected than public.
You're not supposed to call it with some made-up array of types that you think you could support this way.
I know you can still produce nonsense when deriving but I somehow consider deriving from a class as more "advanced" usage than just instantiating it.

roji replied:
Agreed, made the change.

Brar commented:
@roji now that #2450 has landed - any plans to merge this soon?

roji commented:
Yes, I'll rebase this on dev very soon and then you can give this a final sign-off, thanks for the reminder.

roji commented:
Rebased on dev (with the nullability), note that I also renamed RangeHandler.FieldTypes to SupportedElementClrTypes for better consistency and clarity.

@YohDeadfall this is waiting on your sign-off.

YohDeadfall commented on src/Npgsql/TypeMapping/ConnectorTypeMapper.cs:
```diff
@@ -380,20 +380,12 @@ void BindRangeType(NpgsqlTypeHandler elementHandler, PostgresRangeType pgRangeTy
-            }
-
-            BindType(rangeHandler, pgRangeType, rangeNpgsqlDbType, null, clrTypes);
+            BindType((NpgsqlTypeHandler)rangeHandler, pgRangeType, rangeNpgsqlDbType, null, clrTypes.ToArray());
```
> Regarding perf, this is one of those places where I think we should prefer design correctness and codebase quality over such a micro-optimization.

But this doesn't look as quality code because of the cast.

roji replied:
What exactly are you proposing? That we go back to the current situation (removing 0faeb5e5d1348e819a6ff0321000bd7928700c73), making RangeHandler<T> extend an abstract non-generic RangeHandler as before, and duplicating the methods from NpgsqlTypeHandler<T> (e.g. WriteObjectWithLength)? That doesn't make sense to me...

roji replied:
I can make the cast a bit nicer with a check and throw, if that's what's bothering you...

YohDeadfall commented on src/Npgsql/TypeMapping/ConnectorTypeMapper.cs:
```diff
@@ -380,20 +380,12 @@ void BindRangeType(NpgsqlTypeHandler elementHandler, PostgresRangeType pgRangeTy
+            // We only want to bind supported range CLR types whose element CLR types are being bound as well.
+            var clrTypes = rangeHandler.SupportedElementClrTypes
+                .Where(t => elementClrTypes.Contains(t.GenericTypeArguments[0]))
+                .ToList();
```
And this one too. You produce a list of supported types (two allocations, one for list and another one for an internal buffer). In line 388 you are doing `ToArray` and introduce the third.

roji replied:
This is initialization code that happens once per physical connection - the impact of allocations here is non-existent. It's very important to distinguish between hot paths and very cold paths, where it's fine to use LINQ and other relatively inefficient code-improving techniques...

roji replied:
On an unrelated note, if we think there's a perf issue with physical connection setup time (which is only an app setup time), we could consider initializing the whole type handler system lazily rather than eagerly, as we do now. However, we haven't seen any evidence (or gotten complaints) that this is the case, and overheads such as networking, auth, PostgreSQL-side process spinup etc. are very likely to make this completely negligible.

roji commented:
@YohDeadfall ping

roji commented:
@YohDeadfall can you please take a look? It would be good to finish this one.

roji commented:
@roji to rebase

YohDeadfall reviewed (approved):
Since we are going to redesign type handlers in 6.0, I have no objections to have the workaround for now.
